internal/resources/datastore/datastore_storage_config.go

```diff
@@ -161,7 +148,20 @@ func (r *Config) mutate(ctx context.Context, tenantControlPlane *kamajiv1alpha1.
 			username = u
 			password = p
 		} else {
-			username = coalesceFn(tenantControlPlane.Status.Storage.Setup.User)
+		// prioritize the username stored in the TenantControlPlane status,
+		// although this is going to be populated by the UpdateTenantControlPlaneStatus handler of the resource datastore-setup:
+		// the default value will be used for fresh new configurations, and preserving a previous one:
+		// this will keep us safe from naming changes cases as occurred with the following commit:
+		// https://github.com/clastix/kamaji/pull/203/commits/09ce38f489cccca72ab728a259bc8fb2cf6e4770
+			if len(tenantControlPlane.Status.Storage.Setup.User) > 0 {
+				username = []byte(tenantControlPlane.Status.Storage.Setup.User)
+			} else if (len(tenantControlPlane.Spec.DataStoreUsername) > 0) {
+				username = []byte(tenantControlPlane.Spec.DataStoreUsername)
+			} else {
+				// The dash character (-) must be replaced with an underscore, PostgreSQL is complaining about it:
+				// https://github.com/clastix/kamaji/issues/328
+				username = []byte(strings.ReplaceAll(fmt.Sprintf("%s_%s", tenantControlPlane.GetNamespace(), tenantControlPlane.GetName()), "-", "_"))
+			}
```

Comment:
What about a switch statement here?

```suggestion
			switch {
			case len(tenantControlPlane.Status.Storage.Setup.User) > 0:
				username = []byte(tenantControlPlane.Status.Storage.Setup.User)
			case len(tenantControlPlane.Spec.DataStoreUsername) > 0:
				username = []byte(tenantControlPlane.Spec.DataStoreUsername)
			default:
				// The dash character (-) must be replaced with an underscore, PostgreSQL is complaining about it:
				// https://github.com/clastix/kamaji/issues/328
				username = []byte(strings.ReplaceAll(fmt.Sprintf("%s_%s", tenantControlPlane.GetNamespace(), tenantControlPlane.GetName()), "-", "_"))
			}
```